### PR TITLE
[rtl] replace Queue with DWBB Fifo

### PR DESCRIPTION
**Pull request by unlsycn**

### CODE REVIEW

File: nix/t1/conversion/sv-to-vcs-simulator.nix [hunk 1]
```diff
@@ -29,6 +30,15 @@ stdenv.mkDerivation rec {
...
     "-timescale=1ns/1ps"
+    # FIXME
+    # "-y"
+    # "${dwbb}/sim_ver"
+    "-v"
+    "${dwbb}/sim_ver/DW_fifoctl_s1_sf.v"
+    "-v"
+    "${dwbb}/sim_ver/DW_fifo_s1_sf.v"
+    "-v"
+    "${dwbb}/sim_ver/DW_ram_r_w_s_dff.v"
```
Comment (unlsycn) [inline nix/t1/conversion/sv-to-vcs-simulator.nix:41]: FIXME

File: build.sc [hunk 2]
```diff
@@ -82,6 +82,36 @@ trait RVDecoderDB extends millbuild.dependencies.rvdecoderdb.common.RVDecoderDBJ
...
+
+  def mainargsIvy = v.mainargs
+
+  override def moduleDeps = super.moduleDeps ++ Seq(dwbb)
+
+  def chiselModule    = Some(chisel)
+  def chiselPluginJar = T(Some(chisel.pluginModule.jar()))
+  def chiselIvy       = None
+  def chiselPluginIvy = None
+}
```
Comment (sequencer) [inline build.sc:113]: stdlib should dep on dwbb rather than extends from it.

File: build.sc [hunk 3]
```diff
@@ -91,6 +121,8 @@ trait T1 extends millbuild.common.T1Module with ScalafmtModule {
   def axi4Module        = axi4
   def hardfloatModule   = hardfloat
   def rvdecoderdbModule = rvdecoderdb
+  def dwbbModule        = dwbb
+  def stdlibModule      = stdlib
```
Comment (sequencer) [inline build.sc:125]: only dep on stdlib

File: build.sc [hunk 4]
```diff
@@ -117,6 +149,8 @@ trait RocketV extends millbuild.common.RocketVModule with ScalafmtModule {
   def riscvOpcodesPath  = T.input(PathRef(os.pwd / "dependencies" / "riscv-opcodes"))
   def hardfloatModule   = hardfloat
   def axi4Module        = axi4
+  def dwbbModule        = dwbb
+  def stdlibModule      = stdlib
```
Comment (sequencer) [inline build.sc:153]: only dep on stdlib

File: common.sc [hunk 5]
```diff
@@ -4,16 +4,17 @@
...
   def chiselModule: Option[ScalaModule]
 
   override def moduleDeps = super.moduleDeps ++ chiselModule
 
   def chiselPluginJar: T[Option[PathRef]]
 
-  override def scalacOptions = T(super.scalacOptions() ++ chiselPluginJar().map(path => s"-Xplugin:${path.path}") ++ Seq("-Ymacro-annotations"))
+  override def scalacOptions = T(
+    super.scalacOptions() ++ chiselPluginJar().map(path => s"-Xplugin:${path.path}") ++ Seq("-Ymacro-annotations")
+  )
```
Comment (sequencer) [inline common.sc:17]: reformat should have an additional commit.

File: common.sc [hunk 6]
```diff
@@ -24,14 +25,16 @@ trait HasChisel
 
   def chiselPluginIvy: Option[Dep]
 
-  override def scalacPluginIvyDeps: T[Agg[Dep]] = T(super.scalacPluginIvyDeps() ++ chiselPluginIvy.map(Agg(_)).getOrElse(Agg.empty[Dep]))
+  override def scalacPluginIvyDeps: T[Agg[Dep]] = T(
+    super.scalacPluginIvyDeps() ++ chiselPluginIvy.map(Agg(_)).getOrElse(Agg.empty[Dep])
+  )
```
Comment (sequencer) [inline common.sc:30]: reformat should have an additional commit.

File: common.sc [hunk 7]
```diff
@@ -24,14 +25,16 @@ trait HasChisel
...
 }
 
 trait HasRVDecoderDB extends ScalaModule {
   def rvdecoderdbModule: ScalaModule
-  def riscvOpcodesPath: T[PathRef]
+  def riscvOpcodesPath:  T[PathRef]
   def moduleDeps = super.moduleDeps ++ Seq(rvdecoderdbModule)
-  def riscvOpcodesTar: T[PathRef] = T {
+  def riscvOpcodesTar:    T[PathRef]      = T {
     val tmpDir = os.temp.dir()
```
Comment (sequencer) [inline common.sc:38]: reformat should have an additional commit.

File: common.sc [hunk 8]
```diff
@@ -41,130 +44,123 @@ trait HasRVDecoderDB extends ScalaModule {
...
+      )
+      .groupBy(_.segments.contains("unratified"))
+      .map {
+        case (true, fs)  => fs.map(os.copy.into(_, tmpDir / "unratified"))
+        case (false, fs) => fs.map(os.copy.into(_, tmpDir))
+      }
     os.proc("tar", "cf", T.dest / "riscv-opcodes.tar", ".").call(tmpDir)
     PathRef(T.dest)
   }
   override def resources: T[Seq[PathRef]] = super.resources() ++ Some(riscvOpcodesTar())
```
Comment (sequencer) [inline common.sc:56]: reformat should have an additional commit.

File: common.sc [hunk 9]
```diff
@@ -41,130 +44,123 @@ trait HasRVDecoderDB extends ScalaModule {
...
 }
 
+trait HasChiselInterface extends HasChisel {
+  def axi4Module:   ScalaModule
+  def dwbbModule:   ScalaModule
+  def stdlibModule: ScalaModule
+  override def moduleDeps = super.moduleDeps ++ Seq(axi4Module, dwbbModule, stdlibModule)
+}
```
Comment (sequencer) [inline common.sc:64]: I don't think we should merge it together.

File: common.sc [hunk 10]
```diff
@@ -41,130 +44,123 @@ trait HasRVDecoderDB extends ScalaModule {
...
-trait T1Module
-  extends ScalaModule
-    with HasChisel
-    with HasRVDecoderDB {
+trait T1Module extends ScalaModule with HasChiselInterface with HasRVDecoderDB {
   def arithmeticModule: ScalaModule
-  def hardfloatModule: ScalaModule
-  def axi4Module: ScalaModule
+  def hardfloatModule:  ScalaModule
   def moduleDeps = super.moduleDeps ++ Seq(arithmeticModule, hardfloatModule, axi4Module)
```
Comment (sequencer) [inline common.sc:70]: reformat should have an additional commit.

File: common.sc [hunk 11]
```diff
@@ -41,130 +44,123 @@ trait HasRVDecoderDB extends ScalaModule {
...
 
 // T1 forked version of RocketCore
-trait RocketModule
-  extends ScalaModule
-    with HasChisel
-    with HasRVDecoderDB {
+trait RocketModule extends ScalaModule with HasChisel with HasRVDecoderDB {
   def rocketchipModule: ScalaModule
   def moduleDeps = super.moduleDeps ++ Seq(rocketchipModule)
 }
```
Comment (sequencer) [inline common.sc:84]: reformat should have an additional commit.

File: common.sc [hunk 12]
```diff
@@ -41,130 +44,123 @@ trait HasRVDecoderDB extends ScalaModule {
...
 // Link T1 example: RocketV+T1
-trait T1RocketModule
-  extends ScalaModule
-    with HasChisel {
+trait T1RocketModule extends ScalaModule with HasChisel {
   def rocketModule: ScalaModule
-  def t1Module: ScalaModule
+  def t1Module:     ScalaModule
 
   def moduleDeps = super.moduleDeps ++ Seq(rocketModule, t1Module)
```
Comment (sequencer) [inline common.sc:98]: reformat should have an additional commit.

File: stdlib/src/Queue.scala [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    val clock = Module.clock
+    val reset = Module.reset
+
+    // TODO: use sync reset for now and wait for t1 to migrate to FixedIOModule
+    // require(reset.typeName == "Bool" || reset.typeName == "AsyncReset")
+    val useAsyncReset = reset.typeName == "AsyncReset"
+
+    if (entries < 2) {
+      return chisel3.util.Queue(enq, entries, pipe, flow)
+    }
```
Comment (sequencer) [inline stdlib/src/Queue.scala:35]: I suggest inline it in the future.

File: stdlib/src/Queue.scala [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+        )
+      )
+    )
+
+    fifo.io.clk   := clock
+    fifo.io.rst_n := ~(reset.asBool)
+
+    fifo.io.diag_n := ~(false.B)
+
+    enq.ready          := !fifo.io.full
```
Comment (sequencer) [inline stdlib/src/Queue.scala:59]: it should optionally couple `deq.ready` based on the `pipe` and `flow`.

File: stdlib/src/Queue.scala [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    fifo.io.push_req_n := ~(enq.fire)
+    fifo.io.data_in    := enq.bits.asUInt
+
+    deq.valid         := !fifo.io.empty
```
Comment (sequencer) [inline stdlib/src/Queue.scala:63]: it should optionally couple `enq.valid` based on the `pipe` and `flow`.

File: stdlib/src/Queue.scala [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+    fifo.io.pop_req_n := ~(deq.fire)
+    deq.bits          := fifo.io.data_out.asTypeOf(deq.bits)
+
+    if (flow) {
+      when(fifo.io.empty) {
+        deq.bits := enq.bits
+        when(enq.valid) { deq.valid := true.B }
+        when(deq.ready) { fifo.io.push_req_n := ~(false.B) }
+      }
+    }
```
Comment (sequencer) [inline stdlib/src/Queue.scala:73]: Don't use `when` here. Please explicitly use Mux to create datapath.

**Review by sequencer — APPROVED** [review]

**Pull request merged**